### DISCUSSION TITLE
bug: tRPC AppRouter is defined as "any" on import in client app (path alias in Typescript used)

### DISCUSSION THREAD

**Discussion by kennysliding** [trpc/trpc#3525] [Q&A]
### Provide environment information

```
  System:
    OS: macOS 13.0.1
    CPU: (8) arm64 Apple M1
    Memory: 107.17 MB / 8.00 GB
    Shell: 5.8.1 - /bin/zsh
  Binaries:
    Node: 18.7.0 - ~/.nvm/versions/node/v18.7.0/bin/node
    npm: 8.18.0 - ~/.nvm/versions/node/v18.7.0/bin/npm
  Browsers:
    Brave Browser: 104.1.42.95
    Chrome: 108.0.5359.124
    Firefox: 107.0
    Safari: 16.1
  npmPackages:
    typescript: ^4.8.2 => 4.9.4
```

### Describe the bug

I am currently using a monorepo with Turborepo to store all my microservices code in a single repository, each with a `package.json` and a `tsconfig.json`, all of them written in nodejs.

in each `tsconfig.json`, the config is identical

```json
{
  "extends": "@tsconfig/node18-strictest-esm",
  "compilerOptions": {
    "outDir": "build",
    "baseUrl": ".",
    "paths": {
      "@/*": ["./*"],
      "@folder-a/*": ["./folder-a/*"], // masked
      "@folder-b/*": ["folder-b/*"] // masked
    }
  },
  "exclude": ["node_modules", "build"]
}
```

I have one service `A` as a tRPC server, and one service `B` as a tRPC client.

When I imported the type definition of `AppRouter` from `A` to `B`, the type definition becomes `any` in `B` when it is properly typed in `A`.

[![Reference pic here.][1]][1]

I researched about it and found similar [issue](https://stackoverflow.com/questions/74185198/typescript-losing-zod-and-trpc-types-across-monorepo-projects-types-result-in), and it seems like path alias may cause this issue.

Is there any solution/alternate cause to this 

  [1]: https://i.stack.imgur.com/kFjKv.png

### Link to reproduction

https://github.com/Nikola-Milovic/monorepo-trpc-issue

### To reproduce

example is taken from [Nikola-Milovic](https://github.com/Nikola-Milovic), but the idea is the same

### Additional information

Issue also posted in stackoverflow [here](https://stackoverflow.com/questions/74958258/trpc-approuter-is-defined-as-any-on-import)

### 👨‍👧‍👦 Contributing

- [ ] 🙋‍♂️ Yes, I'd be down to file a PR fixing this bug!

**Comment by hanayashiki** [comment]
It turns out if `a/tsconfig.json` imports a `.ts` file from 'b/tsconfig.json', `a` will use the config under its `tsconfig.json` to resolve paths, resulting in that it cannot understand what `@/` in `b` stands for. Removing all the use of absolute import in `b` will resolve the problem.

**Reply by kennysliding** [reply]
Thanks for your answer, I understand that this is one of the ways to resolve the issue, but then it will mean that by using tPRC, we have to give up path alias and absolute import. Therefore I seek if there is an alternative method to use both at the same time.

**Reply by hanayashiki** [reply]
> Thanks for your answer, I understand that this is one of the ways to resolve the issue, but then it will mean that by using tPRC, we have to give up path alias and absolute import. Therefore I seek if there is an alternative method to use both at the same time.

Not really. You can still use absolute path at the root tsconfig.json. Setting "baseUrl" to "packages/" and "paths" to '{ "@/*": ["*"] }' then you can refer to your files by "@/package-a/file1", "@/package-b/file2".

**Comment by guoran8** [comment]
read [this](https://github.com/trpc/trpc/discussions/3476) , you can find a solution

**Reply by kennysliding** [reply]
Thanks for the reply, but the question you referred to doesn't seem to have a solution. Can you elaborate more?

**Reply by guoran8** [reply]
<img width="1001" alt="image" src="https://user-images.githubusercontent.com/6728396/217239490-a55609b4-0102-451e-ac53-3bf38fbdba1f.png">
<img width="1001" alt="image" src="https://user-images.githubusercontent.com/6728396/217239591-08ccbba1-0327-4b85-9a5a-fd76151b5863.png">
1.add a tsconfig.json file in your root folder
2.modify your client app tsconfig.json, add `references` field

**Reply by kennysliding** [reply]
I have added a `tsconfig.json` and add the references field pointing to each app as mentioned, but it does not work unfortunately and the type definition is still `any`

**Comment by btmnk** [comment]
I was able to get it working by just referencing the backend/server tsconfig in the frontend tsconfig
```json
{
  "compilerOptions": {
    ...
    "paths": {
      "@I18n": ["./src/i18n"],
      "@I18n/*": ["./src/i18n/*"],
      "@Config": ["./src/config"],
      "@Config/*": ["./src/config/*"],
      "@Constants": ["./src/constants"],
      "@Constants/*": ["./src/constants/*"],
      "@Domain": ["./src/domain"],
      "@Domain/*": ["./src/domain/*"],
      "@Hooks": ["./src/hooks"],
      "@Hooks/*": ["./src/hooks/*"],
      "@Util": ["./src/util"],
      "@Util/*": ["./src/util/*"],
      "@Api": ["./src/api"],
      "@Api/*": ["./src/api/*"],
      "@Components": ["./src/components"],
      "@Components/*": ["./src/components/*"],
      "@Store": ["./src/store"],
      "@Store/*": ["./src/store/*"]
    }
  },
  "include": ["src"],
  "exclude": ["node_modules"],
  "references": [{ "path": "./tsconfig.node.json" }, { "path": "../bot/tsconfig.json" }]
}
```

And marking the backend as a composite:

```json
{
  "compilerOptions": {
   ...
    "rootDir": "./src",
    "baseUrl": "./src",
    "paths": {
      "@Core": ["core"],
      "@Core/*": ["core/*"],
      "@I18n": ["common/i18n"],
      "@Constant": ["common/constants"],
      "@Constant/*": ["common/constants/*"],
      "@Interface": ["common/interface"],
      "@Interface/*": ["common/interface/*"],
      "@Schema": ["common/schema"],
      "@Schema/*": ["common/schema/*"],
      "@Services": ["services"],
      "@Services/*": ["services/*"],
      "@Util": ["common/util"],
      "@Util/*": ["common/util/*"],
      "@Enum": ["common/enum"],
      "@Enum/*": ["common/enum/*"],
      "@Errors": ["common/errors"],
      "@Errors/*": ["common/errors/*"],
      "@Actions": ["actions"],
      "@Actions/*": ["actions/*"],
      "@Repository": ["repository"],
      "@Repository/*": ["repository/*"]
    },
    "composite": true
  },
  "include": ["./src/**/*.ts"],
  "exclude": ["node_modules"]
}
```

Maybe this helps? Not entirely sure if it's the same problem but I too had issues when using path aliases before referencing the backend.
